zeroex/orderwatch/order_watcher.go

```diff
@@ -410,69 +432,89 @@ func (w *Watcher) setupEventWatcher() {
 						}
+						for _, order := range orders {
+							orderWithTxHashes, ok := hashToOrderWithTxHashes[order.Hash]
+							if !ok {
+								hashToOrderWithTxHashes[order.Hash] = &OrderWithTxHashes{
+									Order: order,
+									TxHashes: map[common.Hash]interface{}{
+										log.TxHash: struct{}{},
+									},
+								}
+							} else {
+								orderWithTxHashes.TxHashes[log.TxHash] = struct{}{}
+							}
+						}
 					}
 				}
+				w.generateOrderEventsIfChanged(hashToOrderWithTxHashes)
 			}
 		}
-
 	}()
 }
 
-func (w *Watcher) findOrderAndGenerateOrderEvents(orderHash common.Hash, txHash common.Hash) {
+func (w *Watcher) findOrderAndGenerateOrderEvents(orderHash common.Hash) (*meshdb.Order) {
 	order := meshdb.Order{}
 	err := w.meshDB.Orders.FindByID(orderHash.Bytes(), &order)
 	if err != nil {
 		if _, ok := err.(db.NotFoundError); ok {
-			return // We will receive events from orders we aren't actively tracking
+			// short-circuit. We expect to receive events from orders we aren't actively tracking
+			return nil
 		}
 		logger.WithFields(logger.Fields{
 			"error":     err.Error(),
 			"orderHash": orderHash,
 		}).Warning("Unexpected error using FindByID for order")
+		return nil
 	}
-	w.generateOrderEventsIfChanged([]*meshdb.Order{&order}, txHash)
+	return &order
 }
 
-func (w *Watcher) findOrdersAndGenerateOrderEvents(makerAddress, tokenAddress common.Address, tokenID *big.Int, txHash common.Hash) {
+func (w *Watcher) findOrdersAndGenerateOrderEvents(makerAddress, tokenAddress common.Address, tokenID *big.Int) []*meshdb.Order {
 	orders, err := w.meshDB.FindOrdersByMakerAddressTokenAddressAndTokenID(makerAddress, tokenAddress, tokenID)
 	if err != nil {
 		logger.WithFields(logger.Fields{
 			"error": err.Error(),
 		}).Panic("unexpected query error encountered")
 	}
-	w.generateOrderEventsIfChanged(orders, txHash)
+	return orders
 }
 
-func (w *Watcher) generateOrderEventsIfChanged(orders []*meshdb.Order, txHash common.Hash) {
+func (w *Watcher) generateOrderEventsIfChanged(hashToOrderWithTxHashes map[common.Hash]*OrderWithTxHashes) {
 	signedOrders := []*zeroex.SignedOrder{}
-	orderHashToOrder := map[common.Hash]*meshdb.Order{}
-	for _, order := range orders {
+	for _, orderWithTxHashes := range hashToOrderWithTxHashes {
+		order := orderWithTxHashes.Order
 		if order.IsRemoved && time.Since(order.LastUpdated) > permanentlyDeleteAfter {
 			w.permanentlyDeleteOrder(order)
 			continue
 		}
 		signedOrders = append(signedOrders, order.SignedOrder)
-		orderHashToOrder[order.Hash] = order
+	}
+	if len(signedOrders) == 0 {
+		return // Noop
 	}
 	validationResults := w.orderValidator.BatchValidate(signedOrders)
 
 	orderEvents := []*zeroex.OrderEvent{}
 	for _, acceptedOrderInfo := range validationResults.Accepted {
-		order := orderHashToOrder[acceptedOrderInfo.OrderHash]
+		orderWithTxHashes := hashToOrderWithTxHashes[acceptedOrderInfo.OrderHash]
+		txHashes := []common.Hash{}
```

Comment:
If we know the length of `txHashes` ahead of time, we should initialize using `make`:

```suggestion
		txHashes := make([]common.Hash{}, len(orderWithTxHashes.TxHashes))
```


Then:

```go
for i, txHash := range orderWithTxHashes.TxHashes {
    txHashes[i] = txHash
}
```